### PR TITLE
Add option to opt out of smaller font size for inlay hints.

### LINKED ISSUE #6883: [VSCode] Feature request: allow modifyable font size for type hints
Currently type hints/annotations are in a slightly smaller size than regular text. The setting for this is seemingly fixed at 2 (?) sizes smaller than the Editor: Font Size (editor.fontSize) setting in VSC settings. This leads to the following text being slightly offset from lines above and below, as can be seen in the screenshot. Please let us detach the sizes or have an option for how different they can be.

image

### PR DESCRIPTION
**Pull request by jmederosalvarado**
As requested on issue #6883 this PR provides an option for users to opt out of the smaller font size for inlay hints. Part of #6883.

### CODE REVIEW

File: editors/code/src/inlay_hints.ts [hunk 1]
```diff
@@ -39,11 +45,7 @@ export function activateInlayHints(ctx: Ctx) {
     maybeUpdater.onConfigChange().catch(console.error);
 }
 
-const typeHints = createHintStyle("type");
-const paramHints = createHintStyle("parameter");
-const chainingHints = createHintStyle("chaining");
-
-function createHintStyle(hintKind: "type" | "parameter" | "chaining") {
+function createHintStyle(ctx: Ctx, hintKind: "type" | "parameter" | "chaining"): InlayHintStyle {
```
Comment (SomeoneToIgnore) [inline editors/code/src/inlay_hints.ts:48]: > ctx: Ctx

Feels a bit odd to drag the whole `Ctx` into the function to use a single param out of it.
I propose to pass a single, properly named boolean there instead or a hints' config at most.

**Review by SomeoneToIgnore — COMMENTED** [review]
That looks nice, thank you for coming up with this.

Although I cannot really agree that this PR 
> Closes #6883

As far as I'm understanding the issue's contents and the header, it requests for the ability set an arbitrary style, not just toggle between the two options.

You don't have to change your PR to support that if you don't want' to, but let's replace `Closes` with a `Part of` in that case.

**Comment by SomeoneToIgnore** [comment]
image

Here's some inspiration from VSCode. That actually reminds me there's an ongoing work on upstreaming the inlay hints into VSCode, so you might not put too much efforts into that.

File: editors/code/src/inlay_hints.ts [hunk 2]
```diff
@@ -114,6 +124,17 @@ class HintsUpdater implements Disposable {
         this.syncCacheAndRenderHints();
     }
 
+    updateInlayHintsStyles() {
```
Comment (SomeoneToIgnore) [inline editors/code/src/inlay_hints.ts:127]: Now we change the style unconditionally on every settings change, even if that setting is not related to hints per se.

This causes the hints to blink when changing some totally unrelated setting which does not look good.

**Comment by jmederosalvarado** [comment]
> Although I cannot really agree that this PR 
> > Closes #6883
> 
> As far as I'm understanding the issue's contents and the header, it requests for the ability to set an arbitrary style, not just toggle between the two options.

I agree with you that this doesn't fully address #6883. I've already replaced `Closes` with `Part of`. Thank you.

**Comment by jmederosalvarado** [comment]
> image
> 
> Here's some inspiration from VSCode. That actually reminds me there's an ongoing work on upstreaming the inlay hints into VSCode, so you might not put too much efforts into that.

That looks interesting, I was not aware of that setting. I suppose is up to the extension to support it (I'm not sure about this however, as I'm have not much experience with vscode extensions).

**Comment by SomeoneToIgnore** [comment]
No worries, the screenshot is more like a showcase for the further ideas, not a requirement to do now.

The only concerning thing for me is the flickering that happens during an unrelated setting change.
You can install your code into VSCode (see the docs in the repo for more details), open some rust file with the hints and the settings panel sideways: then, flip some switches on that panel and observe how
* slow the hints are to switch their style (not a big concern, guess you could dispose them somehow better but up to you totally)
* weird the hints to switch for an unrelated setting toggle — that's what I would love to see fixed here: no need to repaint the hints at all if we don't flip the style switch in the settings

And if you could fix that nit about the `ctx` part along the way, you're more than welcome to do so.

**Comment by jmederosalvarado** [comment]
Hello @SomeoneToIgnore. I believe I've addressed the issues you pointed out. Please let me know what you think.

**Comment by SomeoneToIgnore** [comment]
Works like a charm, even the toggle is not quick.
Thanks a lot!

bors r+

**Comment by SomeoneToIgnore** [comment]
changelog feat (first contribution)  Add option to opt out of smaller font size for inlay hints

File: crates/rust-analyzer/src/config.rs [hunk 3]
```diff
@@ -145,6 +145,8 @@ config_data! {
         inlayHints_parameterHints: bool     = "true",
         /// Whether to show inlay type hints for variables.
         inlayHints_typeHints: bool          = "true",
+        /// Whether inlay hints font size should be smaller than editor's font size.
+        inlayHints_smallerHints: bool       = "true",
```
Comment (lnicola) [inline crates/rust-analyzer/src/config.rs:149]: @matklad do you have any opinion about whether to include client- (or Code-) only configuration in `config.rs`?

Comment (matklad) [reply crates/rust-analyzer/src/config.rs:149]: Good catch, fixed in https://github.com/rust-analyzer/rust-analyzer/pull/8683/commits/855f07202896508d288dba77a89df4364670adf5
